src/jit/gentree.h

```diff
@@ -5131,86 +5131,146 @@ struct GenTreeILOffset : public GenTree
+    GenTree* GetRootTree() const
+    {
+        return m_rootTree;
+    }
 
-    Statement* gtNext;
-    Statement* gtPrev;
+    GenTree** GetRootTreePointer()
+    {
+        return &m_rootTree;
+    }
 
-    bool compilerAdded;
+    void SetRootTree(GenTree* rootTree)
+    {
+        m_rootTree = rootTree;
+    }
 
-    Statement* GetNextStmt()
+    GenTree* GetTreeList() const
     {
-        if (gtNext == nullptr)
-        {
-            return nullptr;
-        }
-        else
-        {
-            return gtNext;
-        }
+        return m_treeList;
     }
 
-    Statement* getPrevStmt()
+    void SetTreeList(GenTree* treeHead)
     {
-        if (gtPrev == nullptr)
-        {
-            return nullptr;
-        }
-        else
-        {
-            return gtPrev;
-        }
+        m_treeList = treeHead;
     }
 
-    Statement(GenTree* expr, IL_OFFSETX offset DEBUGARG(unsigned stmtID))
-        : gtStmtExpr(expr)
-        , gtStmtList(nullptr)
-        , gtInlineContext(nullptr)
-        , gtStmtILoffsx(offset)
-#ifdef DEBUG
-        , gtStmtLastILoffs(BAD_IL_OFFSET)
-        , m_stmtID(stmtID)
-#endif
-        , gtNext(nullptr)
-        , gtPrev(nullptr)
-        , compilerAdded(false)
+    InlineContext* GetInlineContext() const
     {
+        return m_inlineContext;
     }
 
-    bool IsPhiDefnStmt()
+    void SetInlineContext(InlineContext* inlineContext)
     {
-        return gtStmtExpr->IsPhiDefn();
+        m_inlineContext = inlineContext;
     }
 
-    unsigned char GetCostSz() const
+    IL_OFFSETX GetILOffsetX() const
     {
-        return gtStmtExpr->GetCostSz();
+        return m_ILOffsetX;
     }
 
-    unsigned char GetCostEx() const
+    void SetILOffsetX(IL_OFFSETX offsetX)
     {
-        return gtStmtExpr->GetCostEx();
+        m_ILOffsetX = offsetX;
     }
 
 #ifdef DEBUG
+
+    IL_OFFSET GetLastILOffset() const
+    {
+        return m_lastILOffset;
+    }
+
+    void SetLastILOffset(IL_OFFSET lastILOffset)
+    {
+        m_lastILOffset = lastILOffset;
+    }
+
     unsigned GetID() const
     {
         return m_stmtID;
     }
+#endif // DEBUG
+
+    Statement* GetNextStmt() const
+    {
+        return m_next;
+    }
+
+    void SetNextStmt(Statement* nextStmt)
+    {
+        m_next = nextStmt;
+    }
+
+    Statement* GetPrevStmt() const
+    {
+        return m_prev;
+    }
+
+    void SetPrevStmt(Statement* prevStmt)
+    {
+        m_prev = prevStmt;
+    }
+
+    bool IsCompilerAdded() const
+    {
+        return m_compilerAdded;
+    }
+
+    void SetCompilerAdded()
+    {
+        m_compilerAdded = true;
+    }
+
+    bool IsPhiDefnStmt() const
+    {
+        return m_rootTree->IsPhiDefn();
+    }
+
+    unsigned char GetCostSz() const
+    {
+        return m_rootTree->GetCostSz();
+    }
+
+    unsigned char GetCostEx() const
+    {
+        return m_rootTree->GetCostEx();
+    }
+
+private:
+    GenTree* m_rootTree; // The root of the expression tree.
```

Comment:
Please put data members first.

Reply:
Why? Our doc requests public members first https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-jit-coding-conventions.md#13-c-classes  (I know, you can't find many classes that follow that, but why a new class should not).

Reply:
Well, the first thing I see when I read this class is the constructor which initializes some data members. It's not obvious if it initializes them all and if it initializes them in the correct order.

Then I see code that works with data members which I haven't seen yet and I have no idea what their type is.

And then finally I see the data members and I get an idea about what the state of the class is made of, what is the size of the class, if members have a reasonable order to avoid alignment holes, if there may be concurrency issues (not usually the case in the JIT but just as an example).

I don't know about coding guidelines doc but I don't read code sitting on my head.

Reply:
All good points, but I think we should stick with the coding conventions specification. The idea there is that when reading a class, the data shouldn't matter to the consumer -- the function contract should matter most. In this case, you could argue there is too much inline implementation and too few per-function comments explaining this contract, thus exposing too many implementation details. I would generally expect the contract (+ inline implementation, if any) to be too large to fit on a page anyway, so you're already expecting to go back-and-forth between contract/implementation and data (or use multiple windows/views).

Reply:
> The idea there is that when reading a class, the data shouldn't matter to the consumer -- the function contract should matter most. 

That argument is suitable for frameworks/class libraries written by some developers for other developers to consume. The JIT isn't such a framework.